### PR TITLE
Stateless function components in JSX

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
In this PR:
* Support for stateless functional components in JSX (see #5478 for an overview)
* Support for "intrinsic" class and non-class attributes in JSX, e.g. `ref` and `key` (see #5478 again)
* Simplified code around getting built-in types out of the JSX namespace
* Added support for the `@libfiles` test directive so we can test the react.d.ts file (and others) with better fidelity. Use this minimally as it will be much slower than a compact testcase
* Renamed `JsxFlags.ClassElement` to `ValueElement`

There will be a sister PR to this one to DefinitelyTyped to add the new types to react.d.ts

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -203,12 +200,17 @@ namespace ts {
             }
         };
 
+        let jsxElementType: ObjectType;
+        /** Things we lazy load from the JSX namespace */
+        const jsxTypes: {[name: string]: ObjectType} = {};
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `Map<ObjectType>`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -7887,9 +7880,31 @@ namespace ts {
...
-                if (links.jsxFlags & JsxFlags.ClassElement) {
+                if (links.jsxFlags & JsxFlags.ValueElement) {
+                    // Get the element instance type (the result of newing or invoking this tag)
                     const elemInstanceType = getJsxElementInstanceType(node);
 
+                    // Is this is a stateless function component? See if its single signature is
+                    // assignable to the JSX Element Type
+                    const callSignature = getSingleCallSignature(getTypeOfSymbol(sym));
+                    const callReturnType = callSignature && getReturnTypeOfSignature(callSignature);
+                    let paramType = callReturnType && callSignature && (callSignature.parameters.length === 0 ? emptyObjectType : getTypeOfSymbol(callSignature.parameters[0]));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: If `callReturnType` is truthy, is it implied that `callSignature` is truthy as well?

File: src/harness/harness.ts [hunk 3]
```diff
@@ -1115,6 +1117,15 @@ namespace Harness {
                     includeBuiltFiles.push({ unitName: builtFileName, content: normalizeLineEndings(IO.readFile(builtFileName), newLine) });
                 }
 
+                // Files from tests\lib that are requested by "@libFiles"
+                if (options.libFiles) {
+                    ts.forEach(options.libFiles.split(","), filename => {
```
Comment (DanielRosenwasser) [inline src/harness/harness.ts]: `fileName`, also use a `for`..`of` if you don't need a forEach

File: src/compiler/checker.ts (hunk 2 above)
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Yes (`const callReturnType = callSignature && ...`)

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: So you can omit the `&& callSignature` then

**Comment by Shadowfaxenator** [comment]
https://github.com/Microsoft/TypeScript/issues/5675

**Comment by gmosx** [comment]
Will this get included in TypeScript 1.7? Alternatively, any ETA?

**Comment by DanielRosenwasser** [comment]
This is slated for TypeScript 1.8. TypeScript 1.7 should be coming out fairly soon if all goes according to plan.

**Pull request merged**